### PR TITLE
Feature/vote for new epoche

### PR DESCRIPTION
**Pull request by Sorizen**

### CODE REVIEW

File: pala.js [hunk 1]
```diff
@@ -84,6 +88,15 @@ raft.server = (id, peers) => {
   };
 };
 
+const getNewProposerIdxFromServer = (epoch) => epoch % NUM_SERVERS + 1
+
+const getVotesCount = (server) => Object.values(server.voteGranted).reduce((acc, item) => {
+  if(item) {
+    acc += 1
+  }
+  return acc
```
Comment (KlausKidman) [inline pala.js]:
```
.reduce(
  (acc, item) =>  item ? acc + 1 : acc,
  MIN_COUNT_OF_VOTES
)

File: pala.js [hunk 2]
```diff
@@ -94,17 +107,13 @@ const stepDown = (model, server, term) => {
 };
 
 rules.startNewElection = (model, server) => {
-  if ((server.state === SERVER_STATES.follower || server.state === SERVER_STATES.candidate) &&
-      server.electionAlarm <= model.time) {
-    server.electionAlarm = makeElectionAlarm(model.time);
-    server.term += 1;
+  const isLeaderExist = model.servers.some(item => item.state === SERVER_STATES.leader)
+  if ((server.state === SERVER_STATES.follower) &&
```
Comment (KlausKidman) [inline pala.js]:
```
if (server.state === SERVER_STATES.follower &&
  server.electionAlarm <= model.time &&
  !isLeaderExist
) {

File: pala.js [hunk 3]
```diff
@@ -118,22 +127,37 @@ rules.sendRequestVote = (model, server, peer) => {
...
+        server.electionAlarm = util.Inf;
+      }
+    clearServer(model, server)
+    server.term += 1
   }
 };
 
+const clearServer = (model, server) => {
+  server.votedFor = null
+  server.electionAlarm = server.state === SERVER_STATES.leader || server.state === SERVER_STATES.stopped ? 0 : makeElectionAlarm(model.time)
```
Comment (KlausKidman) [inline pala.js]: split into a line

File: pala.js [hunk 4]
```diff
@@ -170,28 +194,19 @@ rules.advanceCommitIndex = (model, server) => {
...
-       server.votedFor === request.from) &&
-      (request.lastLogTerm > logTerm(server.log, server.log.length) ||
-       (request.lastLogTerm === logTerm(server.log, server.log.length) &&
-        request.lastLogIndex >= server.log.length))) {
-    granted = true;
+  if (server.state === SERVER_STATES.candidate) {
     server.votedFor = request.from;
     server.electionAlarm = makeElectionAlarm(model.time);
+    server.peers.forEach(peerId => {
+      server.voteGranted[peerId] =
```
Comment (KlausKidman) [inline pala.js]:
```
const isVoteGranted = server.term === request.term &&
  server.id === request.to &&
  peerId === request.from
if (isVoteGranted) {
  server.voteGranted[peerId] = isVoteGranted
}

File: script.js [hunk 5]
```diff
@@ -88,10 +90,10 @@ playback = function() {
   for (var i = 1; i <= NUM_SERVERS; i += 1) {
       var peers = [];
       for (var j = 1; j <= NUM_SERVERS; j += 1) {
-        if (i != j)
+        if (i !== j)
           peers.push(j);
       }
-      state.current.servers.push(raft.server(i, peers));
+      state.current.servers.push(pala.server(i, peers, i === START_PROPOSER_IDX, i === START_PROPOSER_IDX ? START_PROPOSER_IDX : undefined));
```
Comment (KlausKidman) [inline script.js]: `i === START_PROPOSER_IDX` to const

File: pala.js (hunk 1 above)
Comment (Sorizen) [reply pala.js]: done

File: pala.js (hunk 2 above)
Comment (Sorizen) [reply pala.js]: done

File: pala.js (hunk 3 above)
Comment (Sorizen) [reply pala.js]: done

File: pala.js (hunk 4 above)
Comment (Sorizen) [reply pala.js]: done, but need to talk

File: script.js (hunk 5 above)
Comment (Sorizen) [reply script.js]: done

**Review by KlausKidman — APPROVED** [review]

**Pull request merged**